mppe/crawler.go

```
@@ -0,0 +1,257 @@
+			yearCodes: map[int]int{
+				2011: 303,
+				2012: 304,
+				2013: 305,
+				2014: 306,
+				2015: 307,
+				2016: 308,
+				2017: 346,
+				2018: 409,
+				2019: 449,
+				2020: 508,
+			},
+		},
+		"colaboradores": {
+			category: "valores-percebidos-por-todos-os-colaboradores",
+			yearCodes: map[int]int{
+				2016: 365,
+				2017: 366,
+				2018: 410,
+				2019: 450,
+				2020: 496,
+			},
+		},
+		"exerciciosAnteriores": {
+			category: "verbas-referentes-a-exercicios-anteriores",
+			yearCodes: map[int]int{
+				2016: 348,
+				2017: 349,
+				2018: 411,
+				2019: 461,
+				2020: 509,
+			},
+		},
+		"indenizacoesEOutrosPagamentos": {
+			category: "verbas-indenizatorias-e-outras-remuneracoes-temporarias",
+			yearCodes: map[int]int{
+				2018: 415,
+				2019: 451,
+				2020: 510,
+			},
+		},
+	}
+)
+
+// Crawl download all files related to the MPPE salaries and return their local paths
+func Crawl(outputPath string, month, year int) ([]string, error) {
+	paths := make([]string, 8)
+	pathChannel := make(chan string, 8)
+	var wg sync.WaitGroup
+	for _, member := range members {
+		wg.Add(1)
+		go func(member employeeDescriptor, month, year int) {
+			defer wg.Done()
+			link := fmt.Sprintf("%s%d-%s", baseURL, member.yearCodes[year], member.category)
+			resp, err := http.Get(link)
+			if err != nil {
+				return
+			}
+			defer resp.Body.Close()
+			b, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return
+			}
+			htmlAsString := string(b)
+			pattern := pathResolver(month, year, member.category)
+			fileCode, err := findFileIdentifier(htmlAsString, pattern)
+			if err != nil {
+				return
+			}
+			urlToDownload := fmt.Sprintf("%s%d-%s?download=%s", baseURL, member.yearCodes[year], member.category, fileCode)
+			filePath := fmt.Sprintf("%s/%s-%s-%d.xlsx", outputPath, member.category, fmt.Sprintf("%02d", month), year)
+			desiredFile, err := os.Create(filePath)
+			if err != nil {
+				return
```

Não devemos ignorar erros. Se algo der errado devemos falhar o quanto antes (fail fast) e reportar o erro da forma mais clara. Dessa forma, as pessoas que vão cuidar da execução desse coletor no longo prazo irão conseguir saber o que está acontecendo.

Lembrando que ignorar os erros e baixar o que der pode significar não ter os dados completos. E o pior, nesse caso, talvez nem saibamos o que perdemos. Optamos por não ter que lidar com falhas parciais por enquanto. Simplificamos e tratamos falhas parciais como falhas totais e deixamos para humanos decidirem. Colocar o erro no channel (como você estava fazendo), processar esses erros (criando um string legível), logar esse erro legível e abortar o programa com `os.Exit(1)` é uma boa estratégia.

---

Tranquilo, nesse caso a minha dúvida é se coloco esse ox.Exit(1) dentro do crawler.go mesmo ou no main.go, pela sua fala seria direto on crawler.go mas fiquei na duvida

---

Monta um erro bonitinho e retorna ele para o main.go. Lá você imprime ele na saída padrão e acaba o processo com os.Exit(1)